File: C# Examples/07 Model Creation/PartModel/SolidWorksMacro.cs
using System;
using SolidWorks.Interop.sldworks;
using SolidWorks.Interop.swconst;
using Microsoft.VisualBasic;
using System.Linq;

namespace PartModel
{
    public partial class SolidWorksMacro
    {
        public void Main()
        {
            
            ModelDoc2 swDoc = null;
            PartDoc swPart = null;
            DrawingDoc swDrawing = null;
            AssemblyDoc swAssembly = null;
            bool boolstatus = false;
            int longstatus = 0;
            int longwarnings = 0;
            // 
            // New Document
            double swSheetWidth;
            swSheetWidth = 0;
            double swSheetHeight;
            swSheetHeight = 0;
            //swDoc = ((ModelDoc2)(swApp.NewDocument("C:\\...Path Required...\\Part.prtdot", 0, swSheetWidth, swSheetHeight)));
            swDoc = (ModelDoc2)swApp.NewPart();
            swPart = (PartDoc)swDoc;
            //disable graphics updating for this part
            ModelView mv = (ModelView)swDoc.ActiveView;
            mv.EnableGraphicsUpdate = false;

            //store the user's setting
            bool usersSetting = swApp.GetUserPreferenceToggle((int)swUserPreferenceToggle_e.swInputDimValOnCreate);
            swApp.SetUserPreferenceToggle((int)swUserPreferenceToggle_e.swInputDimValOnCreate, false);

            //create sketch     
            double length;
            double xLoc;
            double yLoc;
            string userLength;
            userLength = Interaction.InputBox("Enter the line length in meters:");
            if (!IsNumeric(userLength))
            {
                System.Windows.Forms.MessageBox.Show
                    ("Please enter only values in meters.", "PartModel", 
                    System.Windows.Forms.MessageBoxButtons.OK, 
                    System.Windows.Forms.MessageBoxIcon.Exclamation);
                return;
            }
            length = Convert.ToDouble(userLength);
            boolstatus = swDoc.Extension.SelectByID2("Front Plane", "PLANE", 0, 0, 0, false, 0, null, 0);
            swDoc.SketchManager.InsertSketch(true);
            swDoc.ClearSelection2(true);
            SketchSegment skSegment = null;
            skSegment = ((SketchSegment)(swDoc.SketchManager.CreateLine(0D, 0D, 0D, 0D, length, 0D)));
            DisplayDimension myDisplayDim = null;
            yLoc = length / 2;
            xLoc = -yLoc;
            myDisplayDim = ((DisplayDimension)(swDoc.AddDimension2(xLoc, yLoc, 0)));
            Dimension myDimension = null;
            //myDimension = ((Dimension)(swDoc.Parameter("D1@Sketch1")));
            //better technique that does not rely on dimension name
            //get the dimension from the DisplayDimension that was just created
            myDimension = (Dimension)myDisplayDim.GetDimension();
            myDimension.SystemValue = length;
            xLoc = length / 2;
            yLoc = -xLoc;
            skSegment = ((SketchSegment)(swDoc.SketchManager.CreateLine(0D, 0D, 0D, length, 0D, 0D)));
            myDisplayDim = ((DisplayDimension)(swDoc.AddDimension2(xLoc, yLoc, 0)));
            //myDimension = ((Dimension)(swDoc.Parameter("D2@Sketch1")));
            myDimension = (Dimension)myDisplayDim.GetDimension();
            myDimension.SystemValue = length;
            // 
            // Named View
            swDoc.ShowNamedView2("*Trimetric", 8);

            //create extrusion
            Feature myFeature = null;
            myFeature = ((Feature)(swDoc.FeatureManager.FeatureExtrusionThin2(true, false, false, 0, 0, length, length, false, false, false, false, 0.017453292519943334D, 0.017453292519943334D, false, false, false, false, true, 0.01D, 0.01D, 0.01D, 0, 0, false, 0.005D, true, true, 0, 0, false)));
            //swDoc.ISelectionManager.EnableContourSelection = false;

            //sketch rectangle
            boolstatus = swDoc.Extension.SelectByRay(0.052833363855540938D, 0, 0.040705551627127079D, -0.40003602677931249D, -0.51503807491002407D, -0.75809429405028372D, 0.00097004279600570629D, 2, false, 0, 0);
            swDoc.SketchManager.InsertSketch(true);
            Array vSkLines = null;
            vSkLines = ((Array)(swDoc.SketchManager.CreateCornerRectangle(length, -0.08474418720157928D, 0, 0.043267859458580915D, -0.022578298536018337D, 0)));
            myFeature = ((Feature)(swDoc.FeatureManager.FeatureCut4(true, false, false, 1, 0, 0.01D, 0.01D, false, false, false, false, 0.017453292519943334D, 0.017453292519943334D, false, false, false, false, false, true, true, true, true, false, 0, 0, false, false)));            

            //select face for hole wizard
            boolstatus = swDoc.Extension.SelectByRay(0, length /2, 0.0487889876318377D, -0.40003602677931249D, -0.51503807491002407D, -0.75809429405028372D, 0.00097004279600570629D, 2, false, 0, 0);
            // 
            // Hole Wizard
            object swHoleFeature = null;
            swHoleFeature = swDoc.FeatureManager.HoleWizard5(1, 1, 36, "M10", 
                1, 0.011D, 0.01D, -1, 0.02D, 1.5707963267948966D, 0, 1, 0, 0,
                0, 0, 0, -1, -1, -1, "", false, true, true, true, true, false);
            //extra code removed

            //add fillet
            swDoc.ClearSelection2(true);
            boolstatus = swDoc.Extension.SelectByRay(0.00038426606806751806D, -0.00029846389747945068D, 
                0.01833916875017394D, -0.40003602677931249D, -0.51503807491002407D, 
                -0.75809429405028372D, 0.00097004279600570629D, 1, false, 1, 0);
            SimpleFilletFeatureData2 swFeatData = null;
            swFeatData = ((SimpleFilletFeatureData2)(swDoc.FeatureManager.CreateDefinition((int)swFeatureNameID_e.swFmFillet)));
            // 
            swFeatData.Initialize((int)swSimpleFilletType_e.swConstRadiusFillet);
            // 
            //object swEdge = null;
            //object edgesArray[] = null;
            //swEdge = ((object)(swDoc.ISelectionManager.GetSelectedObject6(1, 1)));
            //Variant edgesVar;
            //swFeatData.Edges = edgesVar;
            //// 
            //swFeatData.AsymmetricFillet = false;
            //swFeatData.DefaultRadius = 0.01D;
            //swFeatData.ConicTypeForCrossSectionProfile = swFeatureFilletCircular;
            //swFeatData.CurvatureContinuous = false;
            //swFeatData.ConstantWidth = 0.01D;
            //swFeatData.IsMultipleRadius = false;
            //swFeatData.OverflowType = swFilletOverFlowType_Default;
            // 
            Feature swFeature = null;
            swFeature = ((Feature)(swDoc.FeatureManager.CreateFeature(swFeatData)));

            //reset the user's setting
            swApp.SetUserPreferenceToggle((int)swUserPreferenceToggle_e.swInputDimValOnCreate, usersSetting);
            mv.EnableGraphicsUpdate = true;
            swDoc.GraphicsRedraw2();
            return;
        }

        //helper function to check string for numeric (double) value
        public bool IsNumeric(string value)
        {
            Double num = 0;
            bool isDouble = false;

            // Check for empty string.
            if (string.IsNullOrEmpty(value))
            {
                return false;
            }

            isDouble = Double.TryParse(value, out num);

            return isDouble;
        }

        // The SldWorks swApp variable is pre-assigned for you.
        public SldWorks swApp;

    }
}

File: 07 Model Creation/PartModel/SolidWorksMacro.vb
Imports SolidWorks.Interop.sldworks
Imports SolidWorks.Interop.swconst
Imports System.Runtime.InteropServices
Imports System


Partial Class SolidWorksMacro
    Public Sub main()

        Dim swDoc As ModelDoc2 = Nothing
        Dim swPart As PartDoc = Nothing
        Dim swDrawing As DrawingDoc = Nothing
        Dim swAssembly As AssemblyDoc = Nothing
        Dim boolstatus As Boolean = False
        Dim longstatus As Integer = 0
        Dim longwarnings As Integer = 0
        '
        'New Document
        Dim swSheetWidth As Double
        swSheetWidth = 0
        Dim swSheetHeight As Double
        swSheetHeight = 0
        swDoc = swApp.NewPart 'CType(swApp.NewDocument("C:\REPLACE WITH A REAL PATH\Templates\Part.prtdot", 0, swSheetWidth, swSheetHeight), ModelDoc2)
        swDoc.ActiveView.enablegraphicsupdate = False
        swPart = swDoc

        'Store the user's setting
        Dim usersSetting As Boolean
        usersSetting = swApp.GetUserPreferenceToggle _
        (swUserPreferenceToggle_e.swInputDimValOnCreate)
        swApp.SetUserPreferenceToggle _
        (swUserPreferenceToggle_e.swInputDimValOnCreate, False)

        'create sketch
        Dim length As Double
        Dim xLoc As Double
        Dim yLoc As Double
        Dim userLength As String
        userLength = InputBox("Enter the line length in meters:")
        If Not IsNumeric(userLength) Then
            MsgBox("Please enter only values in meters.", MsgBoxStyle.Exclamation)
            Exit Sub
        End If

        length = CType(userLength, Double)

        boolstatus = swDoc.Extension.SelectByID2("Front Plane", "PLANE", 0, 0, 0,
        False, 0, Nothing, 0)
        swDoc.SketchManager.InsertSketch(True)
        swDoc.ClearSelection2(True)
        Dim skSegment As SketchSegment = Nothing
        skSegment = swDoc.SketchManager.CreateLine(0, 0, 0, 0, length, 0)
        yLoc = length / 2
        xLoc = -(length / 2)
        Dim myDisplayDim As DisplayDimension = Nothing
        myDisplayDim = swDoc.AddDimension2(xLoc, yLoc, 0)
        Dim myDimension As Dimension = Nothing
        'myDimension = swDoc.Parameter("D1@Sketch1")
        'more generic technique that does not rely on dimension names
        'get the dimension directly from the created display dimension
        myDimension = myDisplayDim.GetDimension()
        myDimension.SystemValue = length
        skSegment = swDoc.SketchManager.CreateLine(0, 0, 0, length, 0, 0)
        xLoc = length / 2
        yLoc = -(length / 2)
        myDisplayDim = swDoc.AddDimension2(xLoc, yLoc, 0)
        'myDimension = swDoc.Parameter("D2@Sketch1")
        myDimension = myDisplayDim.GetDimension()
        myDimension.SystemValue = length

        'Restore the user's setting
        swApp.SetUserPreferenceToggle _
        (swUserPreferenceToggle_e.swInputDimValOnCreate, usersSetting)

        '
        'Named View
        swDoc.ShowNamedView2("*Trimetric", 8)

        Dim myFeature As Feature = Nothing
        myFeature = CType(swDoc.FeatureManager.FeatureExtrusionThin2(True,
        False, False, 0, 0, 0.1R, 0.01R, False, False, False, False,
        0.017453292519943334R, 0.017453292519943334R, False, False,
        False, False, True, 0.01R, 0.01R, 0.01R, 0, 0, False, 0.005R,
        True, True, 0, 0, False), Feature)
        swDoc.ISelectionManager.EnableContourSelection = False
        swDoc.ActiveView.EnableGraphicsUpdate = False
        swDoc.ViewZoomtofit2()

        swDoc.ClearSelection2(True)
        boolstatus = swDoc.Extension.SelectByRay(0.047617750872973374R, 0, 0.046577984656096305R, -0.40003602677931249R, -0.51503807491002407R, -0.75809429405028372R, 0.0014057224606580827R, 2, False, 0, 0)
        swDoc.SketchManager.InsertSketch(True)
        swDoc.ClearSelection2(True)
        Dim vSkLines As Array = Nothing
        vSkLines = CType(swDoc.SketchManager.CreateCornerRectangle(length, -0.0728368268352142R, 0, 0.0525843427129189R, -0.028526944014799938R, 0), Array)
        myFeature = CType(swDoc.FeatureManager.FeatureCut4(True, False, False, 1, 0, 0.01R, 0.01R, False, False, False, False, 0.017453292519943334R, 0.017453292519943334R, False, False, False, False, False, True, True, True, True, False, 0, 0, False, False), Feature)
        swDoc.ISelectionManager.EnableContourSelection = False
        boolstatus = swDoc.Extension.SelectByRay(0, 0.04986486960530101R, 0.049359275006281678R, -0.40003602677931249R, -0.51503807491002407R, -0.75809429405028372R, 0.0014057224606580827R, 2, False, 0, 0)
        '
        'Hole Wizard
        Dim swHoleFeature As Object = Nothing
        swHoleFeature = swDoc.FeatureManager.HoleWizard5(1, 1, 36, "M10", 1, 0.0105R, 0.01R, -1, 0.02R, 1.5707963267948966R, 0, 0, 0, 0, 0, 0, 0, -1, -1, -1, "", False, True, True, True, True, False)

        'create fillet
        swDoc.ClearSelection2(True)
        boolstatus = swDoc.Extension.SelectByRay(0.0005820422746865006R, -0.00045207896334886755R, 0.0420341416409542R, -0.40003602677931249R, -0.51503807491002407R, -0.75809429405028372R, 0.00088016786570743407R, 1, False, 1, 0)
        Dim swFeatData As SimpleFilletFeatureData2 = Nothing
        swFeatData = CType(swDoc.FeatureManager.CreateDefinition(swFeatureNameID_e.swFmFillet), SimpleFilletFeatureData2)
        '
        swFeatData.Initialize(swSimpleFilletType_e.swConstRadiusFillet)
        '
        'Dim swEdge As Object = Nothing
        'Dim edgesArray(0) As Object = Nothing
        'swEdge = CType(swDoc.ISelectionManager.GetSelectedObject6(1, 1),Object)
        'Dim edgesVar As [Variant]
        'swFeatData.Edges = edgesVar
        '
        'swFeatData.AsymmetricFillet = false
        swFeatData.DefaultRadius = 0.01R
        'swFeatData.ConicTypeForCrossSectionProfile = swFeatureFilletCircular
        'swFeatData.CurvatureContinuous = false
        'swFeatData.ConstantWidth = 0.01R
        'swFeatData.IsMultipleRadius = false
        'swFeatData.OverflowType = swFilletOverFlowType_Default
        '
        Dim swFeature As Feature = Nothing
        swFeature = CType(swDoc.FeatureManager.CreateFeature(swFeatData), Feature)
        swDoc.ActiveView.EnableGraphicsUpdate = True

    End Sub
    ''' <summary>
    ''' The SldWorks swApp variable is pre-assigned for you.
    ''' </summary>
    Public swApp As SldWorks
End Class
